File: apps/frontend/src/features/freshness/sla.ts
import type { SLAStatus } from '@/types/freshness'

export const SLA_LABELS: Record<SLAStatus, string> = {
  ok: 'Até 12h',
  warning_12_24: '12h a 24h',
  warning_24_48: '24h a 48h',
  warning_48_7d: '48h a 7d',
  warning_7d_1m: '7d a 1m',
  stale: '>1 mês',
}

export const SLA_TEXT_COLOR: Record<SLAStatus, string> = {
  ok: 'text-status-ok',
  warning_12_24: 'text-status-ok',
  warning_24_48: 'text-status-warn',
  warning_48_7d: 'text-status-warn',
  warning_7d_1m: 'text-status-error',
  stale: 'text-status-error',
}

export const SLA_SHORT_LABELS: Record<SLAStatus, string> = {
  ok: '≤12h',
  warning_12_24: '12-24h',
  warning_24_48: '24-48h',
  warning_48_7d: '48h-7d',
  warning_7d_1m: '7d-1m',
  stale: '>1m',
}

export const SLA_ORDER: SLAStatus[] = [
  'ok',
  'warning_12_24',
  'warning_24_48',
  'warning_48_7d',
  'warning_7d_1m',
  'stale',
]

File: apps/backend/src/observability_hub/core/sla.py
"""Classificação de SLA de freshness (limiares em horas desde a última
atualização) — vive em core/ porque é usada por dois domínios (freshness,
que já usava isso, e quality, que precisa da mesma classificação pro
score de qualidade). Extraído de domains/freshness/repository.py pra
evitar um domínio importar símbolo privado do outro.
"""

from datetime import UTC, datetime

_SLA_THRESHOLDS_HOURS = [
    (12, "ok"),
    (24, "warning_12_24"),
    (48, "warning_24_48"),
    (168, "warning_48_7d"),
    (720, "warning_7d_1m"),
]


def hours_since(modified: datetime | None) -> float | None:
    if modified is None:
        return None
    return (datetime.now(UTC) - modified).total_seconds() / 3600


def sla_status(hours_since_update: float | None) -> str | None:
    if hours_since_update is None:
        return None
    for threshold, status in _SLA_THRESHOLDS_HOURS:
        if hours_since_update <= threshold:
            return status
    return "stale"
